### ISSUE TITLE
[Core] Ore convert price not available

### ISSUE THREAD

**Issue by ProfessorMinius** [andrew6180/TradeSkillMaster#31]
Saronite/cobalt/titanium ore conversion prices will not show and don't work for custom prices. Here's a screenshot. Eternals and other things work, but ores don't.
![image](https://user-images.githubusercontent.com/25408719/211939123-4dda2985-125f-4303-b78a-fc34d4b42cd8.png)

**Comment by Bananaman** [comment]
@ProfessorMinius I just tested this on Warmane, with the [TradeSkillMaster: Revived](https://github.com/Bananaman/TradeSkillMaster/releases/tag/v2.8.3.666) fork.

![image](https://user-images.githubusercontent.com/38923130/226185967-348a4927-5b7a-4a2f-b0de-400f00c378e3.png)

The market value is calculated correctly, as expected.

There is nothing in TSM's pricing algorithm that would treat ores differently than any other item. TSM doesn't even know what an item is. It just grabs the item ID and then grabs the market value from the AuctionDB (as seen above, the market value is working).

When posting the item, it calculates the automatic pricing correctly:

![image](https://user-images.githubusercontent.com/38923130/226186195-03ca22c9-1ac9-4202-a177-a3a5ae5bb03e.png)


You might be experiencing some random bug in the old TSM, which my new project has fixed. Alternatively, your custom pricing formula is wrong.

Here's my advanced pricing algorithm, which you should put in your "Operations: Auctioning: (Your operation): Post" rules. It automatically calculates reasonable prices for items and refuses to sell things unless you can earn more than 2x the vendor sell value (if it's so close to vendor sell, you should just go vendor it without wasting time and losing deposit costs).

![image](https://user-images.githubusercontent.com/38923130/226186309-839bcc73-7fcb-47be-a39f-ae88569f5d98.png)

- Minimum Price: `max(200% vendorsell, 70% dbmarket)` (When Below: Don't Post)
- Maximum Price: `300% dbmarket` (When Above: Post at Maximum)
- Normal Price: `200% dbmarket`

**Comment by ProfessorMinius** [comment]
@Bananaman the problem isn't the price calculation itself, it's that convert() doesn't work for ores (which if I remember correctly, in the retail version of TSM it should work) i.e. it doesn't return the destroy (prospect) value of the ore.

**Comment by Bananaman** [comment]
@ProfessorMinius IYour screenshot says that you are using `convert(dbmarket)`. I just tried the same, and didn't get any errors, but also didn't get any prospecting value either.

I decided to trace the code a bit...

- `TradeSkillMaster/Core/Prices.lua`: This file parses the `convert()` operator. If you do `convert(dbmarket)` it interprets it as "convert with price-source: dbmarket". It then calls `TSMAPI:GetConvertCost(itemString, extraParam)` where `itemString` is `dbmarket` and the other value is nil, at a cursory glance.
- `TradeSkillMaster/Data/Conversions.lua`: Defines `function TSMAPI:GetConvertCost(targetItem, priceSource)`. A quick glance at that code shows that it expects an actual item string (what item to convert), and optionally a price source in the 2nd param which will be used for pricing the converted items (prospected ores for example).

I'll have a quick look again to see if it's truly sending the item string properly, hold on a bit...

**Comment by Bananaman** [comment]
@ProfessorMinius I've found the issue. `convert()` doesn't do what you think it does.

`TSMAPI:GetConvertCost(targetItem, priceSource)` does the following:

- Take the item (let's say Twilight Opal).

```
["item:36927:0:0:0:0:0:0"] = { -- Twilight Opal
		["item:36909:0:0:0:0:0:0"] = {rate=.01, source="prospect"},
		["item:36912:0:0:0:0:0:0"] = {rate=.04, source="prospect"},
		["item:36910:0:0:0:0:0:0"] = {rate=.04, source="prospect"},
	},
```

- Tell you what you can prospect that from. In the list above you can see that you can prospect it from Saronite Ore (ID 36912) at a chance of 0.04x per saronite ore.

- Look up the values of all the ores/items that can be prospected into that gem (Twilight Opal). Twilight Opals come from prospecting these ores: 36909, 36912, 36910.

- Find the cheapest ore that can be prospected into a Twilight Opal.

- Now you know the "convert cost": The cost to get 1x twilight opal, on average, when buying the cheapest ore with an amount of ores sufficient to usually guarantee that you get 1x opal. (So 1/0.04 = buying 25 ores on average). Therefore it tells you that to convert to 1x twilight opal you have to buy 25x saronite ore at a cost of X gold.


That's what the convert function does. There's unfortunately no bug here. It's clearly used by various crafting algorithms to figure out the cost of getting reagents.

In short: `convert()` does NOT figure out how much ore is worth if prospected into gems. It tells you how much a GEM is worth when prospected FROM the cheapest ore.

**Comment by Bananaman** [comment]
@ProfessorMinius Furthermore, a really quick glance in `TradeSkillMaster/Core/Prices.lua` shows that it doesn't seem to have any operator to do what you want (finding the "prospect worth of an ore"). This is an old version of TSM, remember. It's more basic than retail TSM.

**Comment by Bananaman** [comment]
Oh and my advice about pricing formulas still stands. Those formulas I provided are great for automatically figuring out prices for all items without being ripped off.

**Comment by ProfessorMinius** [comment]
@Bananaman alright, I understand. Do you know of a way to get the prospect value, like the value you see in the UI? Would be handy when buying ores when having to calculate the profit margin and turnout.

**Comment by Bananaman** [comment]
@ProfessorMinius The tooltip uses the following code:

https://github.com/Bananaman/TradeSkillMaster/blob/9f78dc9097a812a564a966cb2d1d1e181500668d/TradeSkillMaster/TradeSkillMaster.lua#L501-L540

From that we can see that it uses a few functions:

- `local prospectValue = TSM:GetProspectValue(itemString)` exists and takes an item and calculates the average prospecting value of 1x ore. To see if the item was prospectable, do `if prospectValue > 0 then`. To find out the prospecting value of a whole stack, multiply that value by the ore quantity.
- If detailed prospecting breakdown tooltips (listing every item) are enabled, it further goes into detail by fetching "all items that come from prospecting", in the code chunk that says `if TSM.db.profile.detailedDestroyTooltip then`.
- But only the first thing, `GetProspectValue` would be interesting for your purposes.

Next step: See what calls that function:

```
 ~/./a/c/d/b/b/P/d/W/I/AddOns $ rg GetProspectValue TradeSkillMaster*
TradeSkillMaster/TradeSkillMaster.lua
503:		local prospectValue = TSM:GetProspectValue(itemString)
640:function TSM:GetProspectValue(itemString)
```

Answer: No... There is no way in this old TSM version to get the prospect values in the automatic pricing algorithms. It's only used in the tooltip code. Furthermore, the code of "GetProspectValue" is very limited. It cannot define a pricing source. It just assumes "100% dbmarket" as its source.

The way TSM parses the custom pricing algorithm strings (like `max(200% vendorsell, 70% dbmarket)`) is also an utter mess. It was the code I mentioned earlier when investigating what `convert()` does. Basically, it's some of the worst spaghetti code I have seen in my life. There's zero structure, tons of repetition, and it's a miracle that the existing code even works at all.

One thing is for sure: Modifying `convert(dbmarket)` is out of the question. It could have effects that break lots of areas of TSM's code which rely on the old conversion algoithm. The best solution would be introducing a new codeword like `max(200% vendorsell, 70% dbmarket, 100% prospect)` and hooking the "prospect" keyword into `TSM:GetProspectValue(itemString)`. But I am not offering to do that. It would require lots of research, testing and ensuring that nothing else breaks by introducing more codewords. It'd be a few hours of work to do it *properly* to be sure there aren't any new bugs, since TSM's price string parsing code is an utter fucking mess, and it'd require deep research into all areas of TSM that use the price string parser. Their parser is literally one huge run-on code block with the most insane code "logic" I've seen in years. Massive mess. I just came out of a few days of massive TSM rewrites and the last thing I want is another job in one of the messiest areas of TSM yet. ;)

**Comment by ProfessorMinius** [comment]
@Bananaman Ah yea, introducing new sources would be a mess. I saw the parsing code myself like a few months ago while diving into how the addon works. It's a bloody mess lol. Anyways, thank you for the help. I'll be closing this now.

**Comment by Bananaman** [comment]
The insane code I'm mentioning is here:

https://github.com/Bananaman/TradeSkillMaster/blob/tsm-revived/TradeSkillMaster/Core/Prices.lua

It's hundreds of line of utter fucking garbage code. One of the biggest messes I've ever seen. There's no intelligence to it. It even ends by dynamically creating a function from a huge string text-block, which in turn is a massive fucking bloater of a mess too. That's what would have to be researched, analyzed and properly edited to add a new "prospect" feature.

**Comment by Bananaman** [comment]
> It's a bloody mess lol. Anyways, thank you for the help. I'll be closing this now.

That's a good way to describe it. It's definitely in a messy league of its own. :)

I'll be heading into WoW to relax now with the working TSM fork, finally. Been having four days off from the game just working 8 hours a day on the rewrite instead. Can't wait to get back to the fun! :) By the way, do you happen to play on Warmane's Icecrown?

**Comment by ProfessorMinius** [comment]
> By the way, do you happen to play on Warmane's Icecrown?
I actully play on Lordaeron

**Comment by Bananaman** [comment]
@ProfessorMinius Ah that's awesome, for me it was hard to pick between Lordaeron and Icecrown. Both seemed great! I ended up as Gnomezilla on Icecrown's Alliance. I'm a returner to Warmane after 3 years away, so heck if anyone plays there feel free to message me ingame! :D Anyway, take care and have fun, perhaps I see you on Lordaeron someday if I want more challenging PvE content! ;) Have a great day! :)
